### PR TITLE
feat(flattener): allow users to flatten JSON primitive column

### PR DESCRIPTION
**Pull request by Thenkei**
### General

- [ ] Write an explicit title for the Pull Request, following [Conventional Commits specification](https://www.conventionalcommits.org)
- [ ] Test manually the implemented changes
- [ ] Validate the code quality (indentation, syntax, style, simplicity, readability)

### Security

- [ ] Consider the security impact of the changes made

### CODE REVIEW

File: packages/plugin-flattener/src/flatten-column/customization.ts [hunk 1]
```diff
@@ -5,18 +5,13 @@ import type {
...
-} from '@forestadmin/datasource-toolkit';
+import { ColumnType, ConditionTreeFactory, SchemaUtils } from '@forestadmin/datasource-toolkit';
 import hashRecord from 'object-hash';
 
 import { deepUpdateInPlace, getValue, unflattenPathsInPlace } from './helpers';
 
-export function makeField(columnName: string, path: string, schema: ColumnSchema) {
-  const columnType = getValue({ [columnName]: schema.columnType }, path) as ColumnType;
+export function makeField(columnName: string, path: string, baseColumnType: ColumnType) {
+  const columnType = getValue({ [columnName]: baseColumnType }, path) as ColumnType;
```
Comment (realSpok) [inline packages/plugin-flattener/src/flatten-column/customization.ts:14]: nice ! More understandable

File: packages/plugin-flattener/src/flatten-json-column/index.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    throw new Error('options.columnName and options.columnType are required.');
+  }
+
+  if (!collection) throw new Error('This plugin can only be called when customizing collections.');
+
+  collection.use(flattenColumn, {
+    columnName: options.columnName,
+    columnType: options.columnType,
+    readonly: options.readonly,
+    level: options.level,
```
Comment (realSpok) [inline packages/plugin-flattener/src/flatten-json-column/index.ts:45]: We should compute the level from the 'nestedness' of the columnType

Comment (Thenkei) [reply packages/plugin-flattener/src/flatten-json-column/index.ts:45]: There is an example where we don't want all nested props. :/

File: packages/plugin-flattener/src/flatten-json-column/index.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ * @param options.readonly Should the imported fields be read-only?
+ * @param options.removeOriginalColumn Should the original column be removed?
+ */
+export default async function flattenJsonColumn(
+  dataSource: DataSourceCustomizer,
+  collection: CollectionCustomizer,
+  options?: FlattenJsonColumnOptions,
+): Promise<void> {
+  if (!options || !options.columnName) {
+    throw new Error('options.columnName and options.columnType are required.');
```
Comment (realSpok) [inline packages/plugin-flattener/src/flatten-json-column/index.ts:36]: Note everyone is on ts sadly :)

File: packages/plugin-flattener/src/flatten-json-column/index.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import { ColumnType } from '@forestadmin/datasource-toolkit';
+
+import flattenColumn from '../flatten-column';
+
+export type FlattenJsonColumnOptions = {
+  columnName: string;
+  columnType: { [key: string]: ColumnType };
+  level?: number;
+  readonly?: boolean;
+  removeOriginalColumn?: boolean;
```
Comment (realSpok) [inline packages/plugin-flattener/src/flatten-json-column/index.ts:15]: what about `keepOriginalColumn` and remove it by default ?

File: packages/plugin-flattener/src/flatten-json-column/index.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+};
+
+/**
+ * Flatten a JSON column which have a composite type
+ *
+ * @param dataSource The dataSource customizer provided by the agent.
```
Comment (realSpok) [inline packages/plugin-flattener/src/flatten-json-column/index.ts:21]: It's not very clear how it should be used. 
I ended up doing something like that:
```ts
        actors.use(flattenJsonColumn, {
          columnName: 'meta',
          readonly: false,
          level: 2,
          removeOriginalColumn: true,
          columnType: { bio: 'String', sex: 'String', sidejob: { area: 'String', job: 'String' } },
        });
```

File: packages/plugin-flattener/src/flatten-json-column/index.ts (hunk 2 above)
Comment (realSpok) [reply packages/plugin-flattener/src/flatten-json-column/index.ts:45]: https://stackoverflow.com/a/66065316/2118884

File: packages/plugin-flattener/src/flatten-json-column/index.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+ * @param collection The collection customizer instance provided by the agent.
+ * @param options The options of the plugin.
+ * @param options.columnName The name of the column to import.
+ * @param options.columnType The type of the column to import.
+ * @param options.level The maximum level of nested fields to import.
+ * @param options.readonly Should the imported fields be read-only?
```
Comment (realSpok) [inline packages/plugin-flattener/src/flatten-json-column/index.ts:27]: we could indicate that it is readonly false by default

File: packages/plugin-flattener/src/flatten-json-column/index.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+export default async function flattenJsonColumn(
+  dataSource: DataSourceCustomizer,
+  collection: CollectionCustomizer,
+  options?: FlattenJsonColumnOptions,
+): Promise<void> {
+  if (!options || !options.columnName) {
+    throw new Error('options.columnName and options.columnType are required.');
+  }
+
+  if (!collection) throw new Error('This plugin can only be called when customizing collections.');
```
Comment (realSpok) [inline packages/plugin-flattener/src/flatten-json-column/index.ts:39]: shouldn't we also check that dataSource was provided?

File: packages/plugin-flattener/src/flatten-json-column/index.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+  CollectionCustomizer,
+  DataSourceCustomizer,
+} from '@forestadmin/datasource-customizer';
+
+import { ColumnType } from '@forestadmin/datasource-toolkit';
+
+import flattenColumn from '../flatten-column';
+
+export type FlattenJsonColumnOptions = {
+  columnName: string;
```
Comment (realSpok) [inline packages/plugin-flattener/src/flatten-json-column/index.ts:11]:
```suggestion
  originColumnName: string;
```
or
```suggestion
  jsonColumnName: string;
```

Comment (realSpok) [reply packages/plugin-flattener/src/flatten-json-column/index.ts:11]: to lift the ambiguity with the `targets` column names

File: packages/plugin-flattener/src/flatten-json-column/index.ts (hunk 5 above)
Comment (realSpok) [reply packages/plugin-flattener/src/flatten-json-column/index.ts:21]: We could add this kind of usage code snippet to the jsdoc what do think ?

File: packages/plugin-flattener/src/flatten-json-column/index.ts [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  removeOriginalColumn?: boolean;
+};
+
+/**
+ * Flatten a JSON column which have a composite type
+ *
+ * @param dataSource The dataSource customizer provided by the agent.
+ * @param collection The collection customizer instance provided by the agent.
+ * @param options The options of the plugin.
+ * @param options.columnName The name of the column to import.
```
Comment (realSpok) [inline packages/plugin-flattener/src/flatten-json-column/index.ts:24]:
```suggestion
 * @param options.columnName The name of the json column that you want to flatten.
```

File: packages/plugin-flattener/src/flatten-json-column/index.ts (hunk 8 above)
Comment (realSpok) [reply packages/plugin-flattener/src/flatten-json-column/index.ts:11]: actually `baseColumnType` like you used above is perfect

Comment (Thenkei) [reply packages/plugin-flattener/src/flatten-json-column/index.ts:11]: I wanted to keep it close to the original plugin but why not.

File: packages/plugin-flattener/src/flatten-json-column/index.ts (hunk 7 above)
Comment (Thenkei) [reply packages/plugin-flattener/src/flatten-json-column/index.ts:39]: There's no dataSource in this case. 🗡️

File: packages/plugin-flattener/src/flatten-json-column/index.ts (hunk 6 above)
Comment (Thenkei) [reply packages/plugin-flattener/src/flatten-json-column/index.ts:31]: How do you imagine this?

File: packages/plugin-flattener/src/flatten-json-column/index.ts (hunk 2 above)
Comment (Thenkei) [reply packages/plugin-flattener/src/flatten-json-column/index.ts]: Done

**Review by Thenkei — COMMENTED** [review]
🚀

**Pull request merged**